ORM/DataObjectSchema.php

```diff
@@ -367,4 +368,396 @@ public function classForField($candidateClass, $fieldName)  {
+		$this->checkRelationClass($class, $component, $relationClass, 'has_one');
+		return $relationClass;
+	}
+
+	/**
+	 * Return data for a specific belongs_to component.
+	 *
+	 * @param string $class
+	 * @param string $component
+	 * @param bool $classOnly If this is TRUE, than any has_many relationships in the
+	 * form "ClassName.Field" will have the field data stripped off. It defaults to TRUE.
+	 * @return string|null
+	 */
+	public function belongsToComponent($class, $component, $classOnly = true) {
+		$belongsTo = (array)Config::inst()->get($class, 'belongs_to');
+		if(!isset($belongsTo[$component])) {
+			return null;
+		}
+
+		// Remove has_one specifier if given
+		$belongsTo = $belongsTo[$component];
+		$belongsToClass = strtok($belongsTo, '.');
+
+		// Validate
+		$this->checkRelationClass($class, $component, $belongsToClass, 'belongs_to');
+		return $classOnly ? $belongsToClass : $belongsTo;
+	}
+
+	/**
+	 *
+	 * @param string $parentClass Parent class name
+	 * @param string $component ManyMany name
+	 * @param string|array $specification Declaration of many_many relation type
+	 * @return array
+	 */
+	protected function parseManyManyComponent($parentClass, $component, $specification)
+	{
+		// Check if this is many_many_through
+		if (is_array($specification)) {
+			// Validate join, parent and child classes
+			$joinClass = $this->checkManyManyJoinClass($parentClass, $component, $specification);
+			$parentClass = $this->checkManyManyFieldClass($parentClass, $component, $joinClass, $specification, 'from');
+			$joinChildClass = $this->checkManyManyFieldClass($parentClass, $component, $joinClass, $specification, 'to');
+			return [
+				ManyManyThroughList::class,
+				$parentClass,
+				$joinChildClass,
+				$specification['from'] . 'ID',
+				$specification['to'] . 'ID',
+				$joinClass,
+			];
+		}
+
+		// Validate $specification class is valid
+		$this->checkRelationClass($parentClass, $component, $specification, 'many_many');
+
+		// automatic scaffolded many_many table
+		$classTable = $this->tableName($parentClass);
+		$parentField = "{$classTable}ID";
+		if ($parentClass === $specification) {
+			$childField = "ChildID";
+		} else {
+			$candidateTable = $this->tableName($specification);
+			$childField = "{$candidateTable}ID";
+		}
+		$joinTable = "{$classTable}_{$component}";
+		return [
+			ManyManyList::class,
+			$parentClass,
+			$specification,
+			$parentField,
+			$childField,
+			$joinTable,
+		];
+	}
+
+	/**
+	 * Find a many_many on the child class that points back to this many_many
+	 *
+	 * @param string $childClass
+	 * @param string $parentClass
+	 * @return string|null
+	 */
+	protected function getManyManyInverseRelationship($childClass, $parentClass)
+	{
+		$otherManyMany = Config::inst()->get($childClass, 'many_many', Config::UNINHERITED);
+		if (!$otherManyMany) {
+			return null;
+		}
+		foreach ($otherManyMany as $inverseComponentName => $nextClass) {
+			if ($nextClass === $parentClass) {
```

Comment:
Yeah, supporting for what I called "triangular relationships" was flakey, and led to some weird behaviour infering reciprocal relationships.